File: core/src/com/lastwords/components/draw/DrawSystem.java
package com.lastwords.components.draw;

import com.badlogic.ashley.core.*;
import com.badlogic.ashley.utils.ImmutableArray;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.lastwords.components.position.PositionComponent;
import com.lastwords.components.texture.TextureComponent;

public class DrawSystem extends EntitySystem {

    private SpriteBatch spriteBatch;

    private ImmutableArray<Entity> entities;

    private ComponentMapper<PositionComponent> positionMapper
            = ComponentMapper.getFor(PositionComponent.class);

    private ComponentMapper<TextureComponent> textureMapper
            = ComponentMapper.getFor(TextureComponent.class);

    public DrawSystem(SpriteBatch spriteBatch) {
        this.spriteBatch = spriteBatch;
    }

    @Override
    public void addedToEngine(Engine engine) {
        entities = engine
                .getEntitiesFor(Family.all(PositionComponent.class, TextureComponent.class).get());
    }

    @Override
    public void update(float deltaTime) {
        float textureWidth, textureHeight, xTexture, yTexture;
        for (Entity entity: entities) {
            PositionComponent positionComponent = positionMapper.get(entity);
            TextureComponent textureComponent = textureMapper.get(entity);
            if (positionComponent.position != null
                && textureComponent.textureRegion != null) {
                textureWidth = textureComponent.textureRegion.getRegionWidth();
                textureHeight = textureComponent.textureRegion.getRegionHeight();
                xTexture = positionComponent.position.x - textureWidth / 2f;
                yTexture = positionComponent.position.y - textureHeight / 2f;
                spriteBatch
                        .draw(textureComponent.textureRegion, xTexture, yTexture, textureWidth, textureHeight);
            }
        }
    }
}

File: core/src/com/lastwords/ashley/draw/DrawSystem.kt
package com.lastwords.ashley.draw

import com.badlogic.ashley.core.*
import com.badlogic.ashley.utils.ImmutableArray
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.lastwords.ashley.position.PositionComponent
import com.lastwords.ashley.texture.TextureComponent

class DrawSystem(private val spriteBatch: SpriteBatch?) : EntitySystem() {

    private var entities: ImmutableArray<Entity>? = null

    private val positionMapper = ComponentMapper.getFor(PositionComponent::class.java)

    private val textureMapper = ComponentMapper.getFor(TextureComponent::class.java)

    override fun addedToEngine(engine: Engine?) {
        entities = engine!!
                .getEntitiesFor(Family.all(PositionComponent::class.java, TextureComponent::class.java).get())
    }

    override fun update(deltaTime: Float) {
        var textureWidth: Float
        var textureHeight: Float
        var xTexture: Float
        var yTexture: Float
        spriteBatch!!.begin()
        for (entity in entities!!) {
            val positionComponent = positionMapper.get(entity)
            val textureComponent = textureMapper.get(entity)
            val textureRegion = textureComponent.textureRegion
            if (textureRegion != null) {
                textureWidth = textureRegion.regionWidth.toFloat()
                textureHeight = textureRegion.regionHeight.toFloat()
                xTexture = positionComponent.position.x - textureWidth / 2f
                yTexture = positionComponent.position.y - textureHeight / 2f
                spriteBatch
                        .draw(textureRegion, xTexture, yTexture, textureWidth, textureHeight)
            }
        }
        spriteBatch.end()
    }
}
